agent/src/storage/bucket.rs

```diff
@@ -0,0 +1,169 @@
+use aws_sdk_s3 as s3;
+use s3::error::ProvideErrorMetadata;
+use s3::primitives::ByteStream;
+use ssri::Integrity;
+use utils::errors::{ServalError, ServalResult};
+use utils::structs::Manifest;
+
+#[derive(Debug, Clone)]
+pub struct S3Storage {
+    client: s3::Client,
+    bucket: String,
+}
+
+impl S3Storage {
+    pub fn new(bucket_name: &str, config: aws_config::SdkConfig) -> ServalResult<Self> {
+        let client = s3::Client::new(&config);
+
+        Ok(S3Storage {
+            client,
+            bucket: bucket_name.to_string(),
+        })
+    }
+
+    pub async fn data_by_sri(&self, integrity: Integrity) -> ServalResult<ByteStream> {
+        let key = integrity.to_string();
+        let object = self
+            .client
+            .get_object()
+            .bucket(&self.bucket)
+            .key(key)
+            .send()
+            .await?;
+
+        Ok(object.body)
+    }
+
+    /// Checks if the given job type is present in our data store, using the fully-qualified name.
+    pub async fn data_exists_by_key(&self, key: &str) -> ServalResult<bool> {
+        let result = self
+            .client
+            .head_object()
+            .bucket(&self.bucket)
+            .key(key)
+            .send()
+            .await;
+        Ok(result.is_ok())
+    }
+
+    /// Fetch a manifest by its fully-qualified name.
+    pub async fn manifest(&self, fq_name: &str) -> ServalResult<Manifest> {
+        let key = Manifest::make_manifest_key(fq_name);
+        let object = self
+            .client
+            .get_object()
+            .bucket(&self.bucket)
+            .key(key)
+            .send()
+            .await?;
+        let chunks = object.body.collect().await?;
+        let string = String::from_utf8(chunks.into_bytes().to_vec())?;
+        let manifest: Manifest = toml::from_str(&string)?;
+        Ok(manifest)
+    }
+
+    /// Retrieve a list of all Wasm manifests stored in the bucket.
+    pub async fn manifest_names(&self) -> ServalResult<Vec<String>> {
+        // this is the only complicated one
+        todo!()
```

Comment:
Noting the todo; but do we really even want/need this function? When would we ever need a complete enumeration? Seems very expensive at scale.